core/mutator.js

```diff
@@ -456,6 +454,23 @@ Blockly.Mutator.prototype.workspaceChanged_ = function(e) {
   }
 };
 
+/**
+ * Returns the extra state of the given block (either as XML or a JSO, depending
+ * on the block's definition).
+ * @param {!Blockly.BlockSvg} block The block to get the extra state of.
+ * @return {string} A strigified version of the extra state of the given block.
+ */
+Blockly.Mutator.prototype.getExtraBlockState_ = function(block) {
```

Comment:
This function and Blockly.Events.BlockChange.prototype.getExtraBlockState_ appear to be identical (right down to the strigification).  Copied code is generally a sign of sub-optimal code (bloat and harder to maintain, though easier to write).  Is it practical to extract these out somewhere, or have one call the other?

Sometimes the answer is no -- but do take a look at this.

Reply:
Yeah this is defnitely something I copy-pasta'ed.

I think the root cause of this duplication is that the `BlockChange` type should use subtyping instead of a switch statement. Eg we would have a `BlockFieldChange` event, `BlockMutatorChange` event, etc. This would allow us to encapsulate the logic for getting the extra state in the `BlockMutatorChange` event. For example, we could have the event do a "recording" like move events do.

But right now, because mutator events are not their own type, every time a caller wants to construct a `BlockChange` event for a mutation, they have to use this logic to get the extra state so they can pass it to the constructor :/

I could add this logic as a static function on the `Blockly.Events.BlockChange` type (i.e. `Blockly.Events.BlockChange.getExtraBlockState`) but based on [this comment from Christopher](https://github.com/google/blockly/issues/5026#issuecomment-880806386) about the goog.module conversion, I'm not sure if that's a good idea. What's your opinion @NeilFraser ? Leave the duplication, or move to a static method? (both being temporary solutions)

Reply:
Entirely your call.  Your instincts here are better than mine.  Just highlighting the issue.